{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4bcae01c",
   "metadata": {},
   "source": [
    "# 0. Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ca27f83d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import MDAnalysis as mda\n",
    "import pandas as pd\n",
    "from MDAnalysis.analysis import distances\n",
    "from scipy.linalg import expm\n",
    "import os\n",
    "from IPython.display import clear_output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f61bb0",
   "metadata": {},
   "source": [
    "## Data Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e0885af2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "checked incomplete PDB structures: entry 18531/18532\n",
      "finalizing primary dataset...\n"
     ]
    }
   ],
   "source": [
    "print('reading files...')\n",
    "\n",
    "# Read the tab separated file from BioLip\n",
    "df = pd.read_csv('BioLiP.csv')\n",
    "\n",
    "# list all PDB structure files\n",
    "os.system('ls data/PDB_files > all_pdb_files.txt')\n",
    "\n",
    "print('removing entry with unavailable PDB structures...')\n",
    "\n",
    "# filter for only the data PDB structure available\n",
    "avail_list = []\n",
    "f = open('all_pdb_files.txt','r')\n",
    "for line in f:\n",
    "    avail_list.append(line[3:7])\n",
    "    \n",
    "avail = []\n",
    "for pdb_id in df.PDB_ID:\n",
    "    if pdb_id in avail_list:\n",
    "        avail.append(1)\n",
    "    else:\n",
    "        avail.append(0)\n",
    "        \n",
    "# drop data with unavailable structures\n",
    "df['available'] = avail\n",
    "df = df[df.available==1]\n",
    "df = df.reset_index(drop=True)\n",
    "\n",
    "print('get protein sequences and length from FASTA...')\n",
    "\n",
    "# get peptide sequences\n",
    "peptide_seq = []\n",
    "len_peptide_seq = []\n",
    "\n",
    "# Read the FASTA sequence files of all filtered PDB entries\n",
    "# get the peptide sequence and length to the list for each entry\n",
    "# if peptide not found just add X\n",
    "for i in range( len(df) ):\n",
    "    found = 0\n",
    "    pdb_id = df.iloc[i,:].PDB_ID    \n",
    "    # seeking for peptide chain\n",
    "    f = open('data/FASTA_files/%s.fasta'%pdb_id)\n",
    "    for line in f:\n",
    "        if( line.split()[0][0] == '>' and\n",
    "            line.split()[0][-1] == df.iloc[i,:].peptide_chain_ID):\n",
    "            peptide_seq.append(f.readline()[:-1])\n",
    "            found = 1\n",
    "            break\n",
    "    if found == 0:\n",
    "        peptide_seq.append('X')\n",
    "    f.close()\n",
    "    len_peptide_seq.append(len(peptide_seq[-1]))\n",
    "            \n",
    "# add the FASTA sequence column\n",
    "df['peptide_seq'] = peptide_seq\n",
    "df['len_peptide_seq'] = len_peptide_seq\n",
    "\n",
    "# Then determine the length of receptors and add a column into df\n",
    "len_protein_seq = []\n",
    "for i in range( len(df) ):\n",
    "    len_protein_seq.append(len(df.iloc[i,:][\"Receptor sequence\"]))    \n",
    "df['len_protein_seq'] = len_protein_seq\n",
    "\n",
    "# ****filter for only data with 3-30 amino acid peptide****\n",
    "df = df[df.len_peptide_seq <= 30]\n",
    "df = df[df.len_peptide_seq >= 3].reset_index(drop=True)\n",
    "\n",
    "# ****filter for only data with 30-300 amino acid receptor protein****\n",
    "df = df[df.len_protein_seq <= 300]\n",
    "df = df[df.len_protein_seq >= 30].reset_index(drop=True)\n",
    "\n",
    "print('dropping data with unknown sequences...')\n",
    "\n",
    "# drop PDB with incomplete structural data\n",
    "incomplete_entry = []\n",
    "for i in range( len(df) ):\n",
    "    if 'X' in df.iloc[i,:].bind_res_renum:\n",
    "        incomplete_entry.append(i)\n",
    "df = df.drop(incomplete_entry).reset_index(drop=True)\n",
    "\n",
    "# other incomplete entry with missing residues\n",
    "incomplete = ['3gi0']\n",
    "n_entry = len(df)\n",
    "\n",
    "for entry in range( n_entry ):\n",
    "    pdb_id = df.PDB_ID[entry]\n",
    "    os.system('cp data/PDB_files/pdb'+pdb_id+'.ent protein_temp.pdb')\n",
    "\n",
    "    Receptor_chain = df.Receptor_chain[entry]\n",
    "    len_protein_seq = df.len_protein_seq[entry]\n",
    "\n",
    "    # import a pdb file as the universe\n",
    "    PDB = 'protein_temp.pdb'\n",
    "    u = mda.Universe(PDB,PDB)\n",
    "\n",
    "    # select only groups of receptor chain if receptor is longer than 50 aa\n",
    "    # otherwise select the whole protein\n",
    "    #if len(df['Receptor sequence'][entry]) > 50:\n",
    "        #CA = u.select_atoms('protein and name CA and chainID '+str(Receptor_chain))\n",
    "    #else:\n",
    "        #CA = u.select_atoms('protein and name CA and not chainID '+str(peptide_chain_ID))\n",
    "    CA = u.select_atoms('protein and name CA and chainID '+str(Receptor_chain))\n",
    "    CA.write(\"CA.pdb\")\n",
    "\n",
    "    # create the amino acid sequence list\n",
    "    seq = CA.atoms.resnames\n",
    "    \n",
    "    if len_protein_seq != len(seq):\n",
    "        incomplete.append(pdb_id)\n",
    "\n",
    "    # list of atom index with alternate locations (altLoc)\n",
    "    for i in range(len(CA)):\n",
    "        if CA[i].altLoc != '' and CA[i].altLoc != 'A':\n",
    "            incomplete.append(pdb_id)\n",
    "            break\n",
    "    clear_output(wait=True)\n",
    "    print('checked incomplete PDB structures: entry '+str(entry)+'/'+str(n_entry))\n",
    "    \n",
    "incomplete_entry = []\n",
    "for pdb_id in incomplete:\n",
    "    incomplete_entry += list(df[df.PDB_ID==pdb_id].index)\n",
    "\n",
    "df = df.drop(incomplete_entry).reset_index(drop=True)\n",
    "\n",
    "print('finalizing primary dataset...')\n",
    "\n",
    "# drop unused columns \n",
    "# then assign protein-peptide interaction entry NO.  \n",
    "df = df[[\"PDB_ID\",\"Receptor_chain\",\"peptide_chain_ID\",\"bind_res_renum\",\"Receptor sequence\",\"peptide_seq\",\"len_protein_seq\",\"len_peptide_seq\"]]\n",
    "df = df.reset_index().rename(columns={\"index\":\"entry\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7e1ac416",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entry</th>\n",
       "      <th>PDB_ID</th>\n",
       "      <th>Receptor_chain</th>\n",
       "      <th>peptide_chain_ID</th>\n",
       "      <th>bind_res_renum</th>\n",
       "      <th>Receptor sequence</th>\n",
       "      <th>peptide_seq</th>\n",
       "      <th>len_protein_seq</th>\n",
       "      <th>len_peptide_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1a0n</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>Y8 Y10 W36 P51 N53 Y54</td>\n",
       "      <td>VTLFVALYDYEARTEDDLSFHKGEKFQILNSSEGDWWEARSLTTGE...</td>\n",
       "      <td>PPRPLPVAPGSSKT</td>\n",
       "      <td>58</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1a1m</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Y7 Y9 N63 Y74 N77 Y84 R97 Y99 D114 Y123 T143 K...</td>\n",
       "      <td>GSHSMRYFYTAMSRPGRGEPRFIAVGYVDDTQFVRFDSDAASPRTE...</td>\n",
       "      <td>TPYDINQML</td>\n",
       "      <td>278</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1a1n</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>R62 N63 I66 F67 T69 T73 Y74 S77 N80 Y84 R97 Y9...</td>\n",
       "      <td>GSHSMRYFYTAMSRPGRGEPRFIAVGYVDDTQFVRFDSDAASPRTE...</td>\n",
       "      <td>VPLRPMTY</td>\n",
       "      <td>276</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1a1o</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>Y7 R62 N63 I66 T69 N77 I80 Y84 R97 Y99 Y123 T1...</td>\n",
       "      <td>GSHSMRYFYTAMSRPGRGEPRFIAVGYVDDTQFVRFDSDAASPRTE...</td>\n",
       "      <td>KPIVQYDNF</td>\n",
       "      <td>276</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1a1r</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>V1 G3 E4 V5 Q6 I7 V8 S9 T35 I36 A37 P60 G62</td>\n",
       "      <td>VEGEVQIVSTATQTFLATCINGVCWTVYHGAGTRTIASPKGPVIQM...</td>\n",
       "      <td>GSVVIVGRIVLSGKPA</td>\n",
       "      <td>152</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11901</th>\n",
       "      <td>11901</td>\n",
       "      <td>6vrn</td>\n",
       "      <td>E</td>\n",
       "      <td>P</td>\n",
       "      <td>R29 Y30</td>\n",
       "      <td>AGITQSPRHKVTETGTPVTLRCHQTENHRYMYWYRQDPGHGLRLIH...</td>\n",
       "      <td>HMTEVVRHC</td>\n",
       "      <td>242</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11902</th>\n",
       "      <td>11902</td>\n",
       "      <td>6vxy</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>F24 H40 N79 Y131 Q155 S172 C173 Q174 G175 S177...</td>\n",
       "      <td>IVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGI...</td>\n",
       "      <td>GRGTKSIPPIAFPD</td>\n",
       "      <td>223</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11903</th>\n",
       "      <td>11903</td>\n",
       "      <td>6w05</td>\n",
       "      <td>H</td>\n",
       "      <td>P</td>\n",
       "      <td>F31 G32 W51 H52 F58 D98 S99 Y102 D103 H104 S107</td>\n",
       "      <td>VQLVESGGGVVQPGRSLRLSCAASGFTFRNFGMHWVRQTPGKGLEW...</td>\n",
       "      <td>NPNANPNA</td>\n",
       "      <td>224</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11904</th>\n",
       "      <td>11904</td>\n",
       "      <td>6w05</td>\n",
       "      <td>L</td>\n",
       "      <td>P</td>\n",
       "      <td>Y91 N92 N93 G94</td>\n",
       "      <td>QIVMTQSPATVSVSPGERATLSCRASRSVTSKLAWYQQKPGQAPRL...</td>\n",
       "      <td>NPNANPNA</td>\n",
       "      <td>212</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11905</th>\n",
       "      <td>11905</td>\n",
       "      <td>6w3j</td>\n",
       "      <td>B</td>\n",
       "      <td>C</td>\n",
       "      <td>K15 I17 R18 Q19 T21 K22 K23 K96 Y99 F103 K110 ...</td>\n",
       "      <td>TLRSITSPLVAHRLKPIRQKTKKAVVSILDSEEVCVELVKEYASQE...</td>\n",
       "      <td>IDDEMFYDDHLEAYFEQLAIPG</td>\n",
       "      <td>218</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11906 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       entry PDB_ID Receptor_chain peptide_chain_ID  \\\n",
       "0          0   1a0n              B                A   \n",
       "1          1   1a1m              A                C   \n",
       "2          2   1a1n              A                C   \n",
       "3          3   1a1o              A                C   \n",
       "4          4   1a1r              A                C   \n",
       "...      ...    ...            ...              ...   \n",
       "11901  11901   6vrn              E                P   \n",
       "11902  11902   6vxy              A                C   \n",
       "11903  11903   6w05              H                P   \n",
       "11904  11904   6w05              L                P   \n",
       "11905  11905   6w3j              B                C   \n",
       "\n",
       "                                          bind_res_renum  \\\n",
       "0                                 Y8 Y10 W36 P51 N53 Y54   \n",
       "1      Y7 Y9 N63 Y74 N77 Y84 R97 Y99 D114 Y123 T143 K...   \n",
       "2      R62 N63 I66 F67 T69 T73 Y74 S77 N80 Y84 R97 Y9...   \n",
       "3      Y7 R62 N63 I66 T69 N77 I80 Y84 R97 Y99 Y123 T1...   \n",
       "4            V1 G3 E4 V5 Q6 I7 V8 S9 T35 I36 A37 P60 G62   \n",
       "...                                                  ...   \n",
       "11901                                            R29 Y30   \n",
       "11902  F24 H40 N79 Y131 Q155 S172 C173 Q174 G175 S177...   \n",
       "11903    F31 G32 W51 H52 F58 D98 S99 Y102 D103 H104 S107   \n",
       "11904                                    Y91 N92 N93 G94   \n",
       "11905  K15 I17 R18 Q19 T21 K22 K23 K96 Y99 F103 K110 ...   \n",
       "\n",
       "                                       Receptor sequence  \\\n",
       "0      VTLFVALYDYEARTEDDLSFHKGEKFQILNSSEGDWWEARSLTTGE...   \n",
       "1      GSHSMRYFYTAMSRPGRGEPRFIAVGYVDDTQFVRFDSDAASPRTE...   \n",
       "2      GSHSMRYFYTAMSRPGRGEPRFIAVGYVDDTQFVRFDSDAASPRTE...   \n",
       "3      GSHSMRYFYTAMSRPGRGEPRFIAVGYVDDTQFVRFDSDAASPRTE...   \n",
       "4      VEGEVQIVSTATQTFLATCINGVCWTVYHGAGTRTIASPKGPVIQM...   \n",
       "...                                                  ...   \n",
       "11901  AGITQSPRHKVTETGTPVTLRCHQTENHRYMYWYRQDPGHGLRLIH...   \n",
       "11902  IVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGI...   \n",
       "11903  VQLVESGGGVVQPGRSLRLSCAASGFTFRNFGMHWVRQTPGKGLEW...   \n",
       "11904  QIVMTQSPATVSVSPGERATLSCRASRSVTSKLAWYQQKPGQAPRL...   \n",
       "11905  TLRSITSPLVAHRLKPIRQKTKKAVVSILDSEEVCVELVKEYASQE...   \n",
       "\n",
       "                  peptide_seq  len_protein_seq  len_peptide_seq  \n",
       "0              PPRPLPVAPGSSKT               58               14  \n",
       "1                   TPYDINQML              278                9  \n",
       "2                    VPLRPMTY              276                8  \n",
       "3                   KPIVQYDNF              276                9  \n",
       "4            GSVVIVGRIVLSGKPA              152               16  \n",
       "...                       ...              ...              ...  \n",
       "11901               HMTEVVRHC              242                9  \n",
       "11902          GRGTKSIPPIAFPD              223               14  \n",
       "11903                NPNANPNA              224                8  \n",
       "11904                NPNANPNA              212                8  \n",
       "11905  IDDEMFYDDHLEAYFEQLAIPG              218               22  \n",
       "\n",
       "[11906 rows x 9 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "802447f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
